src/integrate/totalvi/script.py

```diff
@@ -0,0 +1,181 @@
+    
+    trained_model_exists = os.path.isdir(par["reference_model_path"]) and ("model.pt" in os.listdir(par["reference_model_path"]))
+    return not trained_model_exists or par["force_retrain"]
+
+
+def map_query_to_reference(mdata_reference: MuData, mdata_query: MuData, adata_query: AnnData) -> Tuple[scvi.model.TOTALVI, AnnData]:
+    """Build model on the provided reference if necessary, and map query to the reference"""
+
+    adata_reference: AnnData = convert_mudata_to_anndata(mdata_reference, rna_modality_key=par["reference_modality"], protein_modality_key=par["reference_proteins_key"],
+                                                         input_layer=par["input_layer"], hvg_var_key=par["var_input"])
+
+    scvi.model.TOTALVI.setup_anndata(
+        adata_reference,
+        batch_key=par["obs_batch"],
+        protein_expression_obsm_key=par["reference_proteins_key"]
+    )
+
+    if is_retraining_model():
+        vae_reference = build_reference_model(adata_reference, max_train_epochs=par["max_epochs"])
+    else:
+        vae_reference = scvi.model.TOTALVI.load(dir_path=par["reference_model_path"], adata=adata_reference)
+
+    adata_query: AnnData = align_proteins_names(adata_reference, mdata_query, adata_query, reference_proteins_key=par["reference_proteins_key"],
+                                       query_proteins_key=par["query_proteins_key"])
+
+    # Reorder genes and pad missing genes with 0s
+    scvi.model.TOTALVI.prepare_query_anndata(adata_query, vae_reference)
+
+    # Train the model for query
+    vae_query = scvi.model.TOTALVI.load_query_data(
+        adata_query,
+        vae_reference
+    )
+    vae_query.train(par["max_query_epochs"], plan_kwargs=dict(weight_decay=par["weight_decay"]))
+
+    return vae_query, adata_query
+
+def main():
+    logger = _setup_logger()
+
+    mdata_query = mudata.read(par["query"].strip())
+    adata_query = convert_mudata_to_anndata(mdata_query, rna_modality_key=par["query_modality"], protein_modality_key=par["query_proteins_key"],
+                                            input_layer=par["input_layer"], hvg_var_key=par["var_input"])
+
+    if par["reference"].endswith(".h5mu"):
+        logger.info("Reading reference")
+        mdata_reference = mudata.read(par["reference"].strip())
+
+        logger.info("Mapping query to the reference")
+        vae_query, adata_query = map_query_to_reference(mdata_reference, mdata_query, adata_query)
+    else:
+        raise ValueError("Incorrect format of reference, please provide a .h5mu file")
+
+    adata_query.uns["integration_method"] = "totalvi"
+
+    logger.info("Getting the latent representation of query")
+    adata_query.obsm[par["obsm_output"]] = vae_query.get_latent_representation()
+
+    mdata_query.mod[par["query_modality"]] = adata_query
+    try:
+        logger.info("Updating mdata")
+        mdata_query.update()
+    except KeyError:
+        logger.error("Key error was thrown during mdata update. Be careful")
```

Comment:
Could you give an example why/when this occurs?

Reply:
This is a mysterious error, which sometimes appears after updating mudata (or calling other methods that cause an update). Weirdly, after the error the object is fixed, so it is possible to work with it further. I described the problem here: https://github.com/scverse/mudata/issues/27

Reply:
I think I prefer to keep this error in. When running this component in a pipeline, the user might not even notice the warning in the logs. If needed, we can always add it back? I want to know if this is an edge-case.